nomad/plan_apply.go

```diff
@@ -368,14 +373,12 @@ func updateAllocTimestamps(allocations []*structs.Allocation, timestamp int64) {
 func (p *planner) asyncPlanWait(indexCh chan<- uint64, future raft.ApplyFuture,
 	result *structs.PlanResult, pending *pendingPlan) {
 	defer metrics.MeasureSince([]string{"nomad", "plan", "apply"}, time.Now())
+	defer close(indexCh)
```

Comment:
Would this alone have fixed the deadlock? I think so and since we always use this index inside a `max(...)` it seems like an always safe operation.

Oof, good lesson about writing defensive code the first time round.

Reply:
You never know if you will have time to come back and make it better in a second pass 🤣

Reply:
> Would this alone have fixed the deadlock?

Yes! That's how we did the initial test and fix. Nil'ing the channel gets us two additional things:
* It's belt-and-suspenders so that we know there's no possible way we're still polling on the channel.
* It locally communicates to the reader that the channel is being discarded, as opposed to having to go read the `asyncPlanWait` code further down (and doing so is super cheap).